control-plane/pkg/prober/composite_prober_test.go

```diff
@@ -224,8 +225,204 @@ func TestCompositeProber(t *testing.T) {
+				wantHttpRequestCountMin.Dec()
+				require.Equal(t, network.ProbeHeaderValue, r.Header.Get(network.ProbeHeaderName))
+				writer.WriteHeader(tc.responseStatusCode)
+			})
+			httpServer := httptest.NewUnstartedServer(httpHandler)
+			httpServer.Start()
+			defer httpServer.Close()
+
+			wantHttpsRequestCountMin := atomic.NewInt64(int64(tc.wantHttpsRequestCountMin))
+			httpsHandler := http.HandlerFunc(func(writer http.ResponseWriter, r *http.Request) {
+				wantHttpsRequestCountMin.Dec()
+				require.Equal(t, network.ProbeHeaderValue, r.Header.Get(network.ProbeHeaderName))
+				writer.WriteHeader(tc.responseStatusCode)
+			})
+			httpsServer := httptest.NewUnstartedServer(httpsHandler)
+			cert, err := tls.X509KeyPair(Crt1, Key1)
+			require.NoError(t, err)
+			httpsServer.TLS = &tls.Config{
+				Certificates: []tls.Certificate{cert},
+			}
+			httpsServer.StartTLS()
+			defer httpsServer.Close()
+
+			for _, p := range tc.pods {
+				_ = podinformer.Get(ctx).Informer().GetStore().Add(p)
+			}
+
+			for _, addr := range tc.addressable.AddressStatus.Addresses {
+				if addr.URL.Scheme == "http" {
+					addr.URL.Host = httpServer.URL
+				} else {
+					addr.URL.Host = httpsServer.URL
+				}
+			}
+
+			var IPsLister IPsLister = func(addressable Addressable) ([]string, error) {
+				pods, err := podinformer.Get(ctx).Lister().List(tc.podsLabelsSelector)
+				if err != nil {
+					return nil, err
+				}
+				ips := make([]string, 0, len(pods))
+				for _, p := range pods {
+					ips = append(ips, p.Status.PodIP)
+				}
+				return ips, nil
+			}
+
+			httpUrl, _ := url.Parse(httpServer.URL)
+			requeueCountMin := atomic.NewInt64(int64(0))
+			prober, err := NewCompositeNoTLS(ctx, httpUrl.Port(), IPsLister, func(key types.NamespacedName) {
+				requeueCountMin.Inc()
+				t.Logf("Requeing prober, requeueCount: %d", requeueCountMin.Load())
+			})
+			require.NoError(t, err)
+
+			probeFunc := func() bool {
+				t.Logf("addressable: %+v", tc.addressable.AddressStatus.Addresses)
+				status := prober.Probe(ctx, tc.addressable, StatusReady)
+				return status == tc.wantStatus
+			}
+
+			require.Eventuallyf(t, probeFunc, 5*time.Second, 250*time.Millisecond, "")
+			require.Eventuallyf(t, func() bool { return wantHttpRequestCountMin.Load() == 0 }, 5*time.Second, 250*time.Millisecond, "got %d, want 0", wantHttpRequestCountMin.Load())
+			require.Eventuallyf(t, func() bool { return wantHttpsRequestCountMin.Load() == 0 }, 5*time.Second, 250*time.Millisecond, "got %d, want 0", wantHttpsRequestCountMin.Load())
+			// As we are expecting some of these tests to never resolve to Ready, the prober will potentially requeue more than the min number of times during the test period
+			// So, we want to test that we got at least as many requeues as the min we expect
+			require.Eventuallyf(t, func() bool { return requeueCountMin.Load() >= int64(tc.wantRequeueCountMin) }, 5*time.Second, 250*time.Millisecond, "got %d, want 0", requeueCountMin.Load())
```

Comment:
Do we know why is this the case and does not happen in the other test?

Will we be able to detect that hard-looping doesn't happen? 

For example, currently, the test won't fail if `requeueCountMin` is like 1k (and so it's hard requeue-ing constantly)

Reply:
Yeah, the other tests are testing expecting it to succeed. We are testing here with `exepectedStatus = StatusReady`, but we don't want it to actually succeed. So, it will keep retrying.